File: raspberry_controller/gpio_driver.py
import os

BASE = 512  # gpiochip0 base en tu sistema

class GPIO:
    def __init__(self, bcm_pin):
        self.bcm_pin = bcm_pin
        self.pin = BASE + bcm_pin
        os.system(f"echo {self.pin} > /sys/class/gpio/export")

    def set_direction(self, direction="out"):
        os.system(f"echo {direction} > /sys/class/gpio/gpio{self.pin}/direction")

    def write(self, value: int):
        os.system(f"echo {value} > /sys/class/gpio/gpio{self.pin}/value")

    def read(self) -> int:
        with open(f"/sys/class/gpio/gpio{self.pin}/value", "r") as f:
            return int(f.read().strip())

    def cleanup(self):
        os.system(f"echo {self.pin} > /sys/class/gpio/unexport")

File: IMAGE_CONFIG/meta-gpio/recipes-libs/gpio/files/gpio_driver.c
#include <stdio.h>
#include <stdlib.h>

#define BASE 512

int gpio_export(int bcm_pin) {
    int pin = BASE + bcm_pin;
    char cmd[64];
    snprintf(cmd, sizeof(cmd), "echo %d > /sys/class/gpio/export", pin);
    return system(cmd);
}

int gpio_unexport(int bcm_pin) {
    int pin = BASE + bcm_pin;
    char cmd[64];
    snprintf(cmd, sizeof(cmd), "echo %d > /sys/class/gpio/unexport", pin);
    return system(cmd);
}

int gpio_set_direction(int bcm_pin, const char* direction) {
    int pin = BASE + bcm_pin;
    char cmd[128];
    snprintf(cmd, sizeof(cmd),
             "echo %s > /sys/class/gpio/gpio%d/direction",
             direction, pin);
    return system(cmd);
}

int gpio_write(int bcm_pin, int value) {
    int pin = BASE + bcm_pin;
    char cmd[128];
    snprintf(cmd, sizeof(cmd),
             "echo %d > /sys/class/gpio/gpio%d/value",
             value, pin);
    return system(cmd);
}

int gpio_read(int bcm_pin) {
    int pin = BASE + bcm_pin;
    char path[128];
    FILE *f;
    int val;
    snprintf(path, sizeof(path), "/sys/class/gpio/gpio%d/value", pin);
    f = fopen(path, "r");
    if (f == NULL) return -1;
    fscanf(f, "%d", &val);
    fclose(f);
    return val;
}
